Add basic torch.hash_tensor op

mikaylagawarecki opened the pull request:
Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction

We don't provide an ordering aware hash function yet because we don't have an easy way to provide a fast cuda kernel for it

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154149



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

mikaylagawarecki committed:
Add basic xor_sum op

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic xor_sum op"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic xor_sum op"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 albanD voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic xor_sum op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"




[ghstack-poisoned]

rgommers commented on torch/_torch_docs.py:
```diff
@@ -5010,6 +5010,63 @@ def merge_dicts(*dicts):
+add_docstr(
+    torch.hash_tensor,
+    r"""
+hash_tensor(input, mode=0) -> Tensor
```
From the `hash_tensor` name, I expected it to return a hash of the tensor contents, something one could find in the `hashlib` stdlib module. From the PR description it has nothing to do with that, so I'd say don't use `hash` in the name then. Without looking at it in very much detail, I get the feeling this is pretty unique/niche functionality and an explicit name like `xor_reduce` would describe better what it aims to provide?

Chillee replied:
I think it is intended to be used as a hash function/checksum in many cases.

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- `xor` reduction is technically only defined over integer types. For floating point types the contract for `hash_tensor` is that the tensor will be `.view(dtype)` as the signed integer type of the same bitwidth before the reduction.
- It might be a desirable property for the hash to always return a consistent dtype for all tensors and *technically* in order to implement a hash with xor we should be able to view the underlying storage as an integer type (e.g. int32) and xor it as such.
- However, when we hash over specific dimensions, if we follow the semantic of `.view(dtype)` this would result in the result being a different shape from what one would expect from a reduction op.
e.g. Let's say we we have a float16 tensor of shape [2, 4] and we want to reduce over the first dimension

```python
>>> t = torch.randn(2, 4, dtype=torch.bfloat16)
>>> t = t.view(torch.int32)
>>> t.shape
(2, 2)
>>> hash(t, dim=0)
(1, 2)
```

Further, the above is already the happy case, there are also a bunch of cases where .view(dtype) would fail.  Given the feedback that it would be nice to hash over dimensions (like our reduction ops), I decided to always use the signed integer type of the same bitwidth.

**Happy to hear feedback that this is a terrible choice and we should instead be writing a custom kernel to do a more implicit `view(dtype)` that's doesn't necessarily do the view on the innermost dimension as well as implicitly padding of dims for cases where .view(dtype) would fail**




[ghstack-poisoned]

mikaylagawarecki commented on torch/_torch_docs.py:
```diff
@@ -5010,6 +5010,63 @@ def merge_dicts(*dicts):
+
+{keepdim_details}
+
+Args:
```
This is the same as the documentation for e.g. `torch.sum` where there is docs for "reduce over everything" as well as "reduce with dimension"

https://github.com/pytorch/pytorch/blob/16361b0c40e586aed02453e012a8cf4f3cba23e0/torch/_torch_docs.py#L10746-L10797

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to multiply/shift via range then reduction with xor.

The tensor is always viewed as int64 (applying padding as necessary) before the reduction so the result is always int64.

I chose int64 rather than uint64 because e.g. a * x + b is not implemented for uint64 in torch on cuda.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction





[ghstack-poisoned]

albanD commented on torch/_torch_docs.py:
```diff
@@ -5018,6 +5018,60 @@ def merge_dicts(*dicts):
+add_docstr(
+    torch.hash_tensor,
+    r"""
+hash_tensor(input, mode=0) -> Tensor
```
```suggestion
hash_tensor(input, *, mode=0) -> Tensor
```

albanD commented on torch/_torch_docs.py:
```diff
@@ -5018,6 +5018,60 @@ def merge_dicts(*dicts):
+    >>> torch.hash_tensor(a)
+    tensor(13822780554648485888, dtype=torch.uint64)
+
+.. function:: hash_tensor(input, dim, keepdim=False, mode=0) -> Tensor
```
```suggestion
.. function:: hash_tensor(input, dim, *, keepdim=False, mode=0) -> Tensor
```

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5867,6 +5867,15 @@
+  variants: function, method
+  structured_delegate: hash_tensor.out
+
+- func: hash_tensor.out(Tensor self, int[1] dim=[], bool keepdim=False, int mode=0, *, Tensor(a!) out) -> Tensor(a!)
```
```suggestion
- func: hash_tensor.out(Tensor self, int[1] dim=[], *, bool keepdim=False, int mode=0, *, Tensor(a!) out) -> Tensor(a!)
```

For kwarg-only binding

albanD commented on aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:
```diff
@@ -154,6 +154,51 @@ struct prod_functor<c10::complex<at::Half>> {
+          a_converter.d = a;
+          b_converter.d = b;
+          result_converter.u = a_converter.u ^ b_converter.u;
+          // return a double, otherwise uint64_t will be cast to double
+          // when accumulating and the result will be wrong
```
Ho why?

mikaylagawarecki replied:
Lets say we have inputs 

a: 3.14159, a_bits: 4614256650576692846
b: 1.61803, b_bits: 4609965778477721196
a ^ b: 9219082337818812418

If we return this as int64_t, the next time it is used as an input, it will be cast to 9219082337818812418.0, which becomes 4890905006165143848 in bits, which would mess up the xor reduction here

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add basic torch.hash_tensor op"

Added `torch.hash_tensor` reduction function with a `mode` argument that defaults to reduction with xor.

- The hash is always uint64.
- Integers will be casted to uint64 before performing the xor_sum reduction
- Floats will be upcasted to double and then bitcasted to uint64 before performing the xor_sum reduction





[ghstack-poisoned]

albanD reviewed (approved):
Sounds good to me!

mikaylagawarecki commented:
@pytorchbot merge
